TheMET/TheMET/TagListView/TagListView.swift

```diff
@@ -0,0 +1,205 @@
+            if restTagsWidth < width {
+                let restTagsFrames:[CGRect] = self.setRestTagsFrames(from: tagViewIndex, to: self.tagViews.count - 1, xSpace: xSpace, ySpace: ySpace, x: x, y: y)
+                result.tagViewFrames.append(contentsOf: restTagsFrames)
+                break throughTags
+            } else {
+                let restTagsAndShowMoreButtonFrames = self.setRestTagsAndShowMoreButtonFrames(from: tagViewIndex, to: self.tagViews.count - 1, xSpace: xSpace, ySpace: ySpace, x: x, y: y)
+                result.tagViewFrames.append(contentsOf: restTagsAndShowMoreButtonFrames.tagViewFrames)
+                result.showMoreButtonFrame = restTagsAndShowMoreButtonFrames.showMoreButton
+                break throughTags
+            }
+        }
+    }
+        self.totalHeight = self.tagViews[0].bounds.height * CGFloat(lineNumber + 1) + ySpace * CGFloat(lineNumber)
+        result.height = self.totalHeight
+        return result
+    }
+    
+    private func updateTags() {
+        self.showMoreTagsButton.removeFromSuperview()
+        for tagView in self.tagViews {
+            tagView.removeFromSuperview()
+        }
+        self.tagViews = []
+        var newTagViews: [TagView] = []
+        for tag in self.tags {
+            var configuration = UIButton.Configuration.plain()
+            configuration.contentInsets = NSDirectionalEdgeInsets(top: 5, leading: 10, bottom: 5, trailing: 10)
+            let newTagView = TagView(configuration: configuration)
+            newTagView.setTitle(tag, for: .normal)
+            newTagViews.append(newTagView)
+            newTagView.translatesAutoresizingMaskIntoConstraints = false
+            newTagView.frame = CGRect(origin: .zero,
+                                      size: newTagView.intrinsicContentSize)
+            newTagView.translatesAutoresizingMaskIntoConstraints = true
+            newTagView.autoresizingMask = [ .flexibleRightMargin, .flexibleBottomMargin ]
+        }
+        self.tagViews = newTagViews
+        self.layoutTagViews()
+    }
+    
+    private func layoutTagViews() {
+        let visibleTags:[CGRect] = self.calculateLayout(for: self.bounds.width).tagViewFrames
+        let visibleTagsCount = visibleTags.count
+    throughTags: for (tagViewIndex, tagView) in tagViews.enumerated() {
+        if tagViewIndex < visibleTagsCount {
+            self.addSubview(tagView)
+            tagView.apply(radius: 15, backgroundColor: UIColor(named: "plum"), fontColor: UIColor(named: "blackberry"), font: NSLocalizedString("san_serif_font", comment: ""), fontSize: 13, buttonTitle: (self.tagViews[tagViewIndex].titleLabel?.text)!)
+            let tagFrameOrigin = self.calculateLayout(for: self.bounds.width).tagViewFrames[tagViewIndex].origin
+            tagView.frame.origin = tagFrameOrigin
+        }
+    }
+        if let showMoreButtonFrame = self.calculateLayout(for: self.bounds.width).showMoreButtonFrame,
```

Comment:
You are executing `self.calculateLayout(...)` 3 times in this method. It doesn't make sense. Please execute it a single time an use the results.

Reply:
fixed